TESTS/events/equeue/main.cpp

```diff
@@ -0,0 +1,832 @@
+    timing_func(&fragment->timing);
+
+    struct fragment *nfragment = reinterpret_cast<struct fragment *>(equeue_alloc(fragment->q, fragment->size));
+    TEST_ASSERT(nfragment);
+
+    *nfragment = *fragment;
+    equeue_event_delay(nfragment, fragment->timing.delay);
+
+    int id = equeue_post(nfragment->q, fragment_func, nfragment);
+    TEST_ASSERT(id);
+}
+
+struct cancel {
+    equeue_t *q;
+    int id;
+};
+
+void cancel_func(void *p)
+{
+    struct cancel *cancel = (struct cancel *)p;
+    equeue_cancel(cancel->q, cancel->id);
+}
+
+struct nest {
+    equeue_t *q;
+    void (*cb)(void *);
+    void *data;
+};
+
+void nest_func(void *p)
+{
+    struct nest *nest = (struct nest *)p;
+    equeue_call(nest->q, nest->cb, nest->data);
+
+    ThisThread::sleep_for(10);
+}
+
+void multithread_thread(void *p)
+{
+    equeue_t *q = reinterpret_cast<equeue_t *>(p);
+    equeue_dispatch(q, -1);
+}
+
+void background_func(void *p, int ms)
+{
+    *(unsigned *)p = ms;
+}
+
+
+// Simple call tests
+static void simple_call_test()
+{
+    equeue_t q;
+    int err = equeue_create(&q, 2048);
+    TEST_ASSERT(!err);
+
+    bool touched = false;
+    equeue_call(&q, simple_func, &touched);
+    equeue_dispatch(&q, 0);
+    TEST_ASSERT(touched);
+
+    equeue_destroy(&q);
+}
+
+static void simple_call_in_test()
+{
+    equeue_t q;
+    int err = equeue_create(&q, 2048);
+    TEST_ASSERT(!err);
+
+    bool touched = false;
+    int id = equeue_call_in(&q, 10, simple_func, &touched);
+    TEST_ASSERT(id);
+
+    equeue_dispatch(&q, 15);
+    TEST_ASSERT(touched);
+
+    equeue_destroy(&q);
+}
+
+static void simple_call_every_test()
+{
+    equeue_t q;
+    int err = equeue_create(&q, 2048);
+    TEST_ASSERT(!err);
+
+    bool touched = false;
+    int id = equeue_call_every(&q, 10, simple_func, &touched);
+    TEST_ASSERT(id);
+
+    equeue_dispatch(&q, 15);
+    TEST_ASSERT(touched);
+
+    equeue_destroy(&q);
+}
+
+static void simple_post_test()
+{
+    equeue_t q;
+    int err = equeue_create(&q, 2048);
+    TEST_ASSERT(!err);
+
+    int touched = false;
+    struct indirect *i = reinterpret_cast<struct indirect *>(equeue_alloc(&q, sizeof(struct indirect)));
+    TEST_ASSERT(i);
+
+    i->touched = &touched;
+    int id = equeue_post(&q, indirect_func, i);
+    TEST_ASSERT(id);
+
+    equeue_dispatch(&q, 0);
+    TEST_ASSERT(*i->touched);
+
+    equeue_destroy(&q);
+}
+
+
+// Misc tests
+static void destructor_test()
+{
+    equeue_t q;
+    int err = equeue_create(&q, 2048);
+    TEST_ASSERT(!err);
+
+    int touched;
+    struct indirect *e;
+    int ids[3];
+
+    touched = 0;
+    for (int i = 0; i < 3; i++) {
+        e = reinterpret_cast<struct indirect *>(equeue_alloc(&q, sizeof(struct indirect)));
+        TEST_ASSERT(e);
+
+        e->touched = &touched;
+        equeue_event_dtor(e, indirect_func);
+        int id = equeue_post(&q, pass_func, e);
+        TEST_ASSERT(id);
+    }
+
+    equeue_dispatch(&q, 0);
+    TEST_ASSERT(touched == 3);
```

Comment:
Use `TEST_ASSERT_EQUAL` whenever possible instead of `TEST_ASSERT`. It generates better error description in case of assert